torch/distributed/_shard/sharding_plan/api.py

```diff
@@ -0,0 +1,63 @@
+import abc
+import torch.nn as nn
+
+from dataclasses import dataclass
+from typing import Dict, List, Optional
+
+from torch.distributed._shard.sharding_spec import ShardingSpec
+
+@dataclass
+class ShardingPlan(object):
+    """
+    Representation of a sharding plan, describes how to shard a module
+    across hosts.
+
+    Args:
+        plan (Dict[str, :class:`torch.distributed._shard.sharding_spec.ShardingSpec`]):
+            a dict describes how to shard the parameters of a module, keyed by the name
+            of parameter to ShardingSpec.
+        output_plan (Dict[str, :class:`torch.distributed._shard.sharding_spec.ShardingSpec`), optional):
+            a dict specifies the layout of a module's output which produces a ShardedTensor,
+            keyed by the name of module to ShardingSpec("" in key means the root module).
+            If specified, outputs are resharded according to the provided sharding specs.
+            Default: `None`
+        collect_local_shards (List[str], optional): a list of string, each element enables
+            a module's sharded output to be gathered as a Tensor to ensure further processsing
```

Comment:
Let's add more details about how we gather this as a Tensor. For example, it might seem to use we actually do comms here. But we should explicitly mention we are just collecting the local shard and returning it.